cartographer/mapping/internal/2d/pose_graph_2d.cc

```diff
@@ -595,6 +605,10 @@ void PoseGraph2D::RunOptimization() {
           old_global_to_new_global * mutable_trajectory_node.global_pose;
     }
   }
+  if (global_slam_result_callback_) {
```

Comment:
I think we should be invoking the callback without holding the mutex. Suppose the user wants to call `GetTrajectoryNodePoses()`, since they're interested in updating their outdated ones. If they do it inside their callback, they will get a deadlock.

Reply:
@ojura That's a great point.I hadn't thought of that. I will update PR accordingly once the RFC is merged.

Reply:
I'd call it in `HandleWorkQueue` (and as ojura pointed out, without holding mutexes).